### PR TITLE
PRO-3574: public beta of inline images in rich text

### PR DESCRIPTION
**Pull request by boutell**
## Summary

This is beta quality support for images in rich text. No one has to enable it yet. We need it for UT and other migration projects that used something similar in A2.

## What are the specific steps to test this change?

See the 

## What kind of change does this PR introduce?
*(Check at least one)*

- [ ] Bug fix
- [X] New feature
- [ ] Refactor
- [ ] Documentation
- [ ] Build-related changes
- [ ] Other

## Make sure the PR fulfills these requirements:

- [X] It includes a) the existing issue ID being resolved, b) a convincing reason for adding this feature, or c) a clear description of the bug it resolves
- [X] The changelog is updated
- [ ] Related documentation has been updated
- [ ] Related tests have been updated

If adding a new feature without an already open issue, it's best to open a **feature request issue** first and wait for approval before working on it.

**Other information:**

### CODE REVIEW

File: modules/@apostrophecms/rich-text-widget/index.js [hunk 1]
```diff
@@ -505,36 +543,15 @@ module.exports = {
...
-
-        return output;
       },
-      async output(_super, req, widget, options, _with) {
-        let i;
-        let content = widget.content || '';
+
+      // Quickly replaces rich text permalink placeholder URLs with
+      // actual, SEO-friendly URLs based on `widget._relatedDocs`
+      linkPermalinks(widget, content) {
```
Comment (boutell) [inline modules/@apostrophecms/rich-text-widget/index.js:550]: Not new, just refactored here and accounting for a shared `_relatedDocs`

File: modules/@apostrophecms/rich-text-widget/index.js [hunk 2]
```diff
@@ -567,10 +585,87 @@ module.exports = {
...
-              content = content.substr(0, right + 1) + self.apos.util.escapeHtml(doc.title) + content.substr(nextLeft);
+              content = content.substring(0, right + 1) + self.apos.util.escapeHtml(doc.title) + content.substring(nextLeft);
+            }
+          }
+        }
+        return content;
+      },
+      // Quickly replaces inline image placeholder URLs with
+      // actual, SEO-friendly URLs based on `widget._relatedDocs`
+      linkImages(widget, content) {
```
Comment (boutell) [inline modules/@apostrophecms/rich-text-widget/index.js:596]: Thought about code reuse between this and linkPermalinks, but the URL styles are different enough that I don't want to over-factor things right now

File: modules/@apostrophecms/rich-text-widget/ui/apos/components/AposTiptapAnchor.vue [hunk 3]
```diff
@@ -80,7 +80,6 @@ export default {
   data () {
     return {
       generation: 1,
-      keepInBounds: true,
```
Comment (boutell) [inline modules/@apostrophecms/rich-text-widget/ui/apos/components/AposTiptapAnchor.vue:83]: Dead code removal

File: modules/@apostrophecms/rich-text-widget/ui/apos/components/AposTiptapAnchor.vue [hunk 4]
```diff
@@ -112,12 +111,6 @@ export default {
       const text = state.doc.textBetween(from, to, '');
       return text !== '';
     },
-    offset() {
```
Comment (boutell) [inline modules/@apostrophecms/rich-text-widget/ui/apos/components/AposTiptapAnchor.vue:115]: Dead code removal

File: modules/@apostrophecms/rich-text-widget/ui/apos/components/AposTiptapImage.vue [hunk 5]
```diff
@@ -0,0 +1,253 @@
+<template>
```
Comment (boutell) [inline modules/@apostrophecms/rich-text-widget/ui/apos/components/AposTiptapImage.vue:1]: Based on Link.vue

File: modules/@apostrophecms/rich-text-widget/ui/apos/components/AposTiptapLink.vue [hunk 6]
```diff
@@ -81,7 +81,6 @@ export default {
     const linkWithType = getOptions().linkWithType;
     return {
       generation: 1,
-      keepInBounds: true,
```
Comment (boutell) [inline modules/@apostrophecms/rich-text-widget/ui/apos/components/AposTiptapLink.vue:84]: Dead code removal

File: modules/@apostrophecms/rich-text-widget/ui/apos/components/AposTiptapLink.vue [hunk 7]
```diff
@@ -173,12 +172,6 @@ export default {
       const text = state.doc.textBetween(from, to, '');
       return text !== '';
     },
-    offset() {
```
Comment (boutell) [inline modules/@apostrophecms/rich-text-widget/ui/apos/components/AposTiptapLink.vue:176]: Dead code removal

File: modules/@apostrophecms/rich-text-widget/ui/apos/tiptap-extensions/Image.js [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+          getAttrs: el => {
+            const img = el.querySelector('img');
+            if (!img) {
+              return false;
+            }
+            const caption = el.querySelector('figcaption');
+            const src = img.getAttribute('src');
+            if (!src) {
+              return false;
+            }
```
Comment (ETLaurent) [inline modules/@apostrophecms/rich-text-widget/ui/apos/tiptap-extensions/Image.js:53]: could be done just before the `querySelector`:

```js
  const img = el.querySelector('img');
  const src = img.getAttribute('src');
  if (!img || !src) {
    return false;
  }
```

File: modules/@apostrophecms/rich-text-widget/ui/apos/tiptap-extensions/Image.js [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+            const routeName = components.pop();
+            if (routeName !== 'src') {
+              return false;
+            }
+            const imageId = components.pop();
+            const style = el.getAttribute('class');
+            if (!imageId) {
+              return false;
+            }
```
Comment (ETLaurent) [inline modules/@apostrophecms/rich-text-widget/ui/apos/tiptap-extensions/Image.js:66]: Can be moved before `const style = el.getAttribute('class');`

File: modules/@apostrophecms/rich-text-widget/ui/apos/components/AposTiptapImage.vue [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+  <div class="apos-image-control">
+    <AposButton
+      type="rich-text"
+      @click="click"
+      :class="{ 'apos-is-active': buttonActive }"
+      :label="tool.label"
+      :icon-only="!!tool.icon"
+      :icon="tool.icon ? tool.icon : false"
```
Comment (ETLaurent) [inline modules/@apostrophecms/rich-text-widget/ui/apos/components/AposTiptapImage.vue:9]:
```suggestion
      :icon="tool.icon || false"
```

File: modules/@apostrophecms/rich-text-widget/ui/apos/components/AposTiptapImage.vue [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+          // stack interchangeably with tiptap's
+          browse: false
+        },
+        ...(getOptions().imageStyles ? [
+          {
+            name: 'style',
+            label: this.$t('apostrophe:style'),
+            type: 'select',
+            choices: getOptions().imageStyles,
+            def: getOptions().imageStyles?.[0].value
```
Comment (ETLaurent) [inline modules/@apostrophecms/rich-text-widget/ui/apos/components/AposTiptapImage.vue:100]: default selected is `null`
 🤔 
image

File: modules/@apostrophecms/rich-text-widget/index.js [hunk 12]
```diff
@@ -567,10 +585,87 @@ module.exports = {
...
+        if (widget._edit) {
+          return content;
+        }
+        // "Why no regexps?" We need to do this as quickly as we can.
+        // indexOf and lastIndexOf are much faster.
+        let i;
+        for (const doc of (widget._relatedDocs || [])) {
+          let offset = 0;
+          while (true) {
```
Comment (ETLaurent) [inline modules/@apostrophecms/rich-text-widget/index.js:605]: Is `while` used for performance here?

**Review by ETLaurent — COMMENTED** [review]
just small comments, tested on testbed ✅

File: modules/@apostrophecms/rich-text-widget/index.js (hunk 12 above)
Comment (boutell) [reply modules/@apostrophecms/rich-text-widget/index.js:605]: Yes. This is brought forward from carefully tuned code in the A2 permalinks module. This is extra work done at read time, and Apostrophe is read-optimized, so getting a little bit micro with the optimizations is warranted, even at some cost in readability.

**Comment by boutell** [comment]
I've addressed the reported issues. Thanks for the nudge to fix the default style choice.

**Review by ETLaurent — APPROVED** [review]

**Pull request merged**